baseplate/ratelimit.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import division
+
+import time
+
+from .context import ContextFactory
+
+
+class RateLimitExceededException(Exception):
+    """This exception gets raised whenever a rate limit is exceeded.
+    """
+    pass
+
+
+class RateLimiterContextFactory(ContextFactory):
+    """RateLimiter context factory
+
+    :param cache_context_factory: An instance of
```

Comment:
are there any other constraints on it that the user should know about?

Reply:
what about having the ratelimit factory wrap a cache context factory instead? then this factory only needs to take one parameter and the ratelimit factories can abstract away what info is needed under their hoods.

Reply:
I think I like this direction, but does this mean we need a class per backend?:
`class RedisRateLimiterContextFactory(RedisContextFactory)` and `class MemcacheRateLimiterContextFactory(MemcacheContextFactory)`?

Reply:
```python
class RateLimiterContextFactory(ContextFactory):
    def __init__(self, backend, allowance, interval):
        ...

class Backend(object):
    def consume(...):
        ...

class RedisBackend(Backend):
    def __init__(self, redis_context_factory, prefix):
        ...

class MemcachedBackend(Backend):
    def __init__(self, memcache_context_factory, prefix):
        ...
```

then the main ratelimiter is completely agnostic to anything about the backends (including key prefixes) and the backend just handles the protocol of talking to the cache/redis/whatever

Reply:
I see. Thanks!

Reply:
Taking a step back, it seems like this is the API we would want:

```
redis_pool = pool_from_config(app_config)
backend = RedisRateLimitBackend(redis_pool)
ratelimiter_factory = RateLimiterContextFactory(backend, allowance, interval)
baseplate.add_to_context('ratelimiter', ratelimiter_factory)
```

So I guess the `Backend` constructors should actually take a cache pool rather than a cache context factory.

Reply:
we do want spans for the cache traffic generated by the backend though, right?

Reply:
Yeah, I think that makes sense. I guess I thought this `pool_from_config` would return a pool of `MonitoredRedisConnection` which would create spans. Maybe I'm missing something though?

Reply:
Oh yeah `MonitoredRedisConnection` gets created by `RedisContextFactory`

Reply:
the base pool is just straight up from the redis library (https://github.com/reddit/baseplate/blob/master/baseplate/context/redis.py#L54) but the context factory adds the baseplatey span stuff. https://github.com/reddit/baseplate/blob/master/baseplate/context/redis.py#L76

Reply:
i'd be ok with the ratelimitbackend doing the context factory wrapping automatically rather than requiring that one be passed in. so it'd look like your snippet above but inside we're adding the contextfactory wrapper.

Reply:
Yeah. That takes us back to your suggested API:
```
redis_pool = pool_from_config(app_config)
redis_factory = RedisContextFactory(redis_pool)
backend = RedisRateLimitBackend(redis_factory)
ratelimiter_factory = RateLimiterContextFactory(backend, allowance, interval)
baseplate.add_to_context('ratelimiter', ratelimiter_factory)
```

Thanks again

Reply:
> i'd be ok with the ratelimitbackend doing the context factory wrapping automatically rather than requiring that one be passed in. so it'd look like your snippet above but inside we're adding the contextfactory wrapper.

I think this might make for a nicer API since it removes a layer of classes